File: trunk/java/SR-RBD/src/ddb/communication/TcpWorker.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package ddb.communication;

import java.io.DataInputStream;

import java.io.IOException;
import java.net.Socket;
import java.util.concurrent.BlockingQueue;

import ddb.Logger;
import ddb.msg.Message;
import ddb.msg.MessageType;
import ddb.msg.InvalidMessageTypeException;

/**
 *
 * @author xeonic
 */
public class TcpWorker implements Runnable
{
    private final static String LOGGING_NAME = "TcpWorker";
    
    protected BlockingQueue<Message> storage;
    protected Socket socket;
    
    public TcpWorker(Socket newSocket, BlockingQueue<Message> queue)
    {
    	storage = queue;
        socket = newSocket;
    }

    public void run()
    {
    	Logger.getInstance().log("New connection from: "
			+ socket.getInetAddress().getHostAddress().toString() + ":"
			+ socket.getPort()
			, LOGGING_NAME, Logger.Level.INFO);
    	do
    	{
    		try
    		{
    			socket.setSoTimeout(0);

    			DataInputStream in = new DataInputStream(socket.getInputStream());
    			int size = in.readInt();
    			int type = in.readInt();
    			byte[] b = new byte[size];

    			Logger.getInstance().log("Size = " + Integer.toString(size),
    					LOGGING_NAME, Logger.Level.INFO);

    			int left;
    			for(left = size; left > 0; )
    				left -= in.read(b, size - left, left);

    			Message m = Message.Unserialize(MessageType.fromInt(type), b);
    			
    			storage.put(m);
    		}
    		catch (IOException ex)
    		{
    			Logger.getInstance().log("Socket exception! " + ex.getMessage(), 
    					LOGGING_NAME,
    					Logger.Level.WARNING);
    			break;
    		} 
    		catch (InterruptedException ex) 
    		{
    			Logger.getInstance().log("InterruptedException " + ex.getMessage(), 
    					LOGGING_NAME,
    					Logger.Level.WARNING);
				break;
			} catch (InvalidMessageTypeException ex) {
				Logger.getInstance().log("InvalidMessageTypeException " + ex.getMessage(), 
    					LOGGING_NAME,
    					Logger.Level.WARNING);
				break;
			}
    		
    	} while(true);
    }
}

File: trunk/C#/RBD/RBD/Communication/TcpWorker.cs
//using System;
//using System.Collections.Generic;
//using System.Linq;
//using System.Text;
//using System.Net;
//using System.Net.Sockets;

//namespace RBD
//{
//    class TcpWorker
//    {
//        const String LOGGING_NAME = "TcpWorker";

        
//        BlockingQueue<Message> storage;
//        TcpClient tcpClient;

//        public TcpWorker(TcpClient newTcpClient, BlockingQueue<Message> newStorage)
//        {
//            tcpClient = newTcpClient;
//            storage = newStorage;
//        }

//        public void run()
//        {
//            IPEndPoint ip = (IPEndPoint)tcpClient.Client.RemoteEndPoint;
//            Logger.getInstance().log("New connection from: " + ip.Address.ToString()
//                                     + ":" + ip.Port.ToString(),
//                                     LOGGING_NAME, Logger.Level.INFO);
//            NetworkStream stream = tcpClient.GetStream();

//            try
//            {
//                while (true)
//                {
//                    Byte[] sizeBuf = new Byte[4];
//                    for (int nread = 0; nread < 4; )
//                        nread += stream.Read(sizeBuf, nread, 4 - nread);
//                    int size = BitConverter.ToInt32(sizeBuf, 0);
//                    size = IPAddress.NetworkToHostOrder(size);

//                    Byte[] typeBuf = new Byte[4];
//                    for (int nread = 0; nread < 4; )
//                        nread += stream.Read(typeBuf, nread, 4 - nread);
//                    int type = BitConverter.ToInt32(typeBuf, 0);
//                    type = IPAddress.NetworkToHostOrder(type);

//                    Logger.getInstance().log("Size = " + size, LOGGING_NAME, Logger.Level.INFO);

//                    Byte[] buffer = new Byte[size];
//                    for (int nread = 0; nread < size; )
//                        nread += stream.Read(buffer, nread, size - nread);

//                    String messageStr = System.Text.Encoding.ASCII.GetString(buffer);
//                    Message m = Message.Unserialize(MessageType.fromInt(type), messageStr, ip.Address);
//                    storage.put(m);

//                    Logger.getInstance().log("Message: [" + message + "]",
//                                             LOGGING_NAME, Logger.Level.INFO);
//                }
//            }
//            catch(System.IO.IOException ex)
//            {
//                Logger.getInstance().log("Socket exception! " + ex.Message, LOGGING_NAME, Logger.Level.WARNING);
//                break;
//            }
//            TcpSender.getInstance().removeNode(ip.Address);
//        }
//    }
//}
